src/Elasticsearch.Net/ConnectionPool/SniffingSortedStickyConnectionPool.cs

```diff
@@ -0,0 +1,145 @@
+
+		/// <inheritdoc/>
+		public bool SupportsPinging => true;
+
+		/// <inheritdoc/>
+		public DateTime LastUpdate { get; protected set; }
+
+		public SniffingSortedStickyConnectionPool(IEnumerable<Uri> uris, Func<IEnumerable<Node>, IEnumerable<Node>> sortMethod, IDateTimeProvider dateTimeProvider = null)
+			: this(uris.Select(uri => new Node(uri)), sortMethod, dateTimeProvider)
+		{
+		}
+
+		public SniffingSortedStickyConnectionPool(IEnumerable<Node> nodes, Func<IEnumerable<Node>, IEnumerable<Node>> sortMethod, IDateTimeProvider dateTimeProvider = null)
+		{
+			nodes.ThrowIfEmpty(nameof(nodes));
+
+			this.DateTimeProvider = dateTimeProvider ?? Elasticsearch.Net.DateTimeProvider.Default;
+
+			var nn = nodes.ToList();
+			var uris = nn.Select(n => n.Uri).ToList();
+			if (uris.Select(u => u.Scheme).Distinct().Count() > 1)
+				throw new ArgumentException("Trying to instantiate a connection pool with mixed URI Schemes");
+
+			this.UsingSsl = uris.Any(uri => uri.Scheme == "https");
+
+			this.InternalNodes = nn
+				.DistinctBy(n => n.Uri)
+				.ToList();
+			this.LastUpdate = this.DateTimeProvider.Now();
+
+			this._sortMethod = sortMethod;
+		}
+
+		protected int GlobalCursor = -1;
+
+		/// <summary>
+		/// Creates a view of all the live nodes preserving their order,
+		/// if there are no live nodes yields a different dead node to try once
+		/// </summary>
+		public IEnumerable<Node> CreateView(Action<AuditEvent, Node> audit = null)
+		{
+			var now = this.DateTimeProvider.Now();
+			var nodes = this.InternalNodes.Where(n => n.IsAlive || n.DeadUntil <= now)
+				.ToList();
+			var count = nodes.Count;
+			Node node;
+
+			if (count == 0)
+			{
+				var globalCursor = Interlocked.Increment(ref this.GlobalCursor);
+				//could not find a suitable node retrying on first node off globalCursor
+				audit?.Invoke(AuditEvent.AllNodesDead, null);
+				node = this.InternalNodes[globalCursor % this.InternalNodes.Count];
+				node.IsResurrected = true;
+				audit?.Invoke(AuditEvent.Resurrection, node);
+				yield return node;
+				yield break;
+			}
+
+			// If the cursor is greater than the default then it's been
+			// set already but we now have a live node so we should reset it
+			if (this.GlobalCursor > -1)
+			{
+				Interlocked.Exchange(ref this.GlobalCursor, -1);
+			}
+
+			var localCursor = 0;
+
+			for (var attempts = 0; attempts < count; attempts++)
+			{
+				node = nodes[localCursor];
+				localCursor = (localCursor + 1) % count;
+				//if this node is not alive or no longer dead mark it as resurrected
+				if (!node.IsAlive)
+				{
+					audit?.Invoke(AuditEvent.Resurrection, node);
+					node.IsResurrected = true;
+				}
+				yield return node;
+			}
+		}
+
+		/// <summary>
+		/// Update the node list and sort it
+		/// </summary>
+		public void Reseed(IEnumerable<Node> nodes)
+		{
+			if (!nodes.HasAny()) return;
+
+			try
+			{
+				this._readerWriter.EnterWriteLock();
+				List<Node> sortedNodes = _sortMethod(nodes).ToList();
```

Comment:
How about passing in an `IComparer<Node>` on the constructor instead of the `Func<IEnumerable<Node>, IEnumerable<Node>>`  to control node ordering? That way Node set can't be filtered here, and it'd allow the desired _server_ > _rack_id_ > _room_ logic in issue #2657 to be implemented.

If I were to take this approach would a including an `IComparer<Node>` implementation which provides that functionality in this PR be acceptable? It would take a `Dictionary<string, string>` in the constructor and compare the values in that to those in `Node.Settings` in order to favour nodes with the greatest number of matching attributes.